### PR TITLE
Remove more integration test port list allocations 

### PR DESCRIPTION
**Pull request by tcsc**
Following on from #13658, this patch removes more (but  unfortunately not _all_) usages of the deprecated, list-based port-allocation scheme.

This patch:

1. Updated the integration test `TeleInstance` fixture to use injected listeners rather than static ports when creating a new proxy node in a cluster,
2. Updates tests affected by (1) to pre-allocate and inject listeners, including handling caching the listener FDs between proxy restarts
3. Removed unnecessary port allocations when creating LoadBalancer fixtures
4. Moved the remaining list-base port allocation functions out of `helpers` and back into `integrations` and made private. These functions should **never** be used by more than one test package concurrently or there is a very high chance of a port collision. Rather than just write that rule down in the comments, I have contained the deprecated code into the affected package made the compiler enforce the rule for us.

See-Also: #12421
See-Also: #13658
See-Also: #14408

### CODE REVIEW

File: integration/helpers/instance.go [hunk 1]
```diff
@@ -1056,12 +1058,12 @@ func (i *TeleInstance) StartProxy(cfg ProxyConfig) (reversetunnel.Server, error)
...
 			ts, ok := re.Payload.(reversetunnel.Server)
 			if ok {
-				return ts, nil
+				return ts, process, nil
 			}
 		}
 	}
 
-	return nil, nil
+	return nil, nil, nil
```
Comment (smallinsky) [inline integration/helpers/instance.go:1066]: Not sure if any test flow reaches this line but  if `process` seems to be available so why not return it ?   
I know that this is legacy code not related to this PR but returning `nil, nil, nil` looks really suspicious.

File: integration/ports.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// Deprecated: Use helpers.NewListener() and friends instead.
+func newPortValue() int {
+	return ports.PopInt()
+}
+
+// newPortStr fetches aport from the pool as a string.
+// Deprecated: Use helpers.NewListener() and friends instead.
+func newPortStr() string {
+	return ports.Pop()
+}
```
Comment (smallinsky) [inline integration/ports.go:48]: Any reason why this packet was moved from `/integration/helpers/ports.go` to `/integration/ports.go` ?

File: integration/helpers/instance.go [hunk 3]
```diff
@@ -967,25 +966,27 @@ func (i *TeleInstance) StartNodeAndProxy(t *testing.T, name string) (sshPort, we
...
 	DisableWebInterface bool
 	// Disable ALPN routing
 	DisableALPNSNIListener bool
+	// FileDescriptors holds FDs to be injected into the Teleport process
+	FileDescriptors []service.FileDescriptor
 }
 
 // StartProxy starts another Proxy Server and connects it to the cluster.
-func (i *TeleInstance) StartProxy(cfg ProxyConfig) (reversetunnel.Server, error) {
+func (i *TeleInstance) StartProxy(cfg ProxyConfig) (reversetunnel.Server, *service.TeleportProcess, error) {
```
Comment (smallinsky) [inline integration/helpers/instance.go:986]: nit: Does returning `*service.TeleportProcess` is really needed ? right now it is used only in one please where 
 the `TeleportProcess can be also obtained by referring to `i.Nodes[len(i.Nodes)-1]` after calling `i.StartProxy`.

File: integration/ports.go (hunk 2 above)
Comment (tcsc) [reply integration/ports.go:48]: See pt 4 in PR comment

**Review by smallinsky — APPROVED** [review]
left some nit comments  otherwise LGTM.

File: integration/helpers/instance.go (hunk 3 above)
Comment (tcsc) [reply integration/helpers/instance.go:986]: Just returning it from the `Start` method seems less invasive to me, but I think reasonable people can differ on this one.

**Review by jakule — APPROVED** [review]

**Review by camscale — COMMENTED** [review]
Just one little nit from a drive-by review.

Also it would be nice to reference https://github.com/gravitational/teleport/pull/13346 in the PR description to understand why this PR is being done.

**Comment by tcsc** [comment]
> Also it would be nice to reference #13346 in the PR description to understand why this PR is being done.

It's actually more closely related to #14408 and #16133, but your point stands.

File: integration/helpers/instance.go (hunk 1 above)
Comment (tcsc) [reply integration/helpers/instance.go]: I've made it an explicit error, as nothing good should happen if control even lands there.

**Pull request merged**